File: codes/zhuohal1/LC92 Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
            return head
        if m > 1:
            new_head = head
            new_head.next = self.reverseBetween(head.next, m-1, n-1)
            return new_head
        else:
            new_next = head.next
            new_head = self.reverseBetween(new_next, 1, n-1)
            new_rest = new_next.next
            new_next.next = head
            head.next = new_rest
            return new_head

File: codes/Yu/LC92_Reverse_Linked_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// 0 ms (100.00%), 34.2 MB (100.00%)
// https://leetcode.com/problems/reverse-linked-list-ii/discuss/300120/JAVA%3A-Simple-and-Concise-solution-with-detailed-explanation
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode start = head;

        for (int i = 0; i < m - 1; i++) {
            prev = prev.next;
            start = start.next;
        }

        // 1->2->3->4->5->6->7 | 1->2->4->3->5->6->7 | 1->2->5->4->3->6->7 |
        // 1->2->6->5->4->3->7 | 1->2->7->6->5->4->3
        for (int i = 0; i < n - m; i++) {
            ListNode curr = start.next;
            if (curr.next == null) {
                start.next = null;
            } else {
                start.next = curr.next;
            }
            curr.next = prev.next;
            prev.next = curr;
        }
        return dummy.next; // return head; is wrong when m = 1
    }
}